<a href="https://colab.research.google.com/github/monicasjsu/Machine_Learning/blob/main/Logistic_Regression_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import io
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import classification_report
from sklearn.metrics import mean_squared_error
from sklearn.utils import resample
from google.colab import drive
import collections

In [ ]:
#Loading Data
drive.mount('/content/drive')
data = pd.read_csv("/content/drive/My Drive/MLSpring2020/shecodes_employee_attrition/HR-Employee-Attrition.csv")


Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [ ]:
data.head()

Age Attrition  ... YearsSinceLastPromotion  YearsWithCurrManager
0   41       Yes  ...                       0                     5
1   49        No  ...                       1                     7
2   37       Yes  ...                       0                     0
3   33        No  ...                       3                     0
4   27        No  ...                       2                     2

[5 rows x 35 columns]

In [ ]:
#Data Preparation
#Checking datatype and other details of dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
Age                         1470 non-null int64
Attrition                   1470 non-null object
BusinessTravel              1470 non-null object
DailyRate                   1470 non-null int64
Department                  1470 non-null object
DistanceFromHome            1470 non-null int64
Education                   1470 non-null int64
EducationField              1470 non-null object
EmployeeCount               1470 non-null int64
EmployeeNumber              1470 non-null int64
EnvironmentSatisfaction     1470 non-null int64
Gender                      1470 non-null object
HourlyRate                  1470 non-null int64
JobInvolvement              1470 non-null int64
JobLevel                    1470 non-null int64
JobRole                     1470 non-null object
JobSatisfaction             1470 non-null int64
MaritalStatus               1470 non-null object
MonthlyIncome         

In [ ]:
#Checking if there are any missing values in dataset
data.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

In [ ]:
#Converting categorical values to numerical
data.Attrition = pd.Categorical(data.Attrition)
data.Attrition = data.Attrition.cat.codes
data.BusinessTravel = pd.Categorical(data.BusinessTravel)
data.BusinessTravel = data.BusinessTravel.cat.codes
data.Department = pd.Categorical(data.Department)
data.Department = data.Department.cat.codes
data.EducationField = pd.Categorical(data.EducationField)
data.EducationField = data.EducationField.cat.codes
data.Gender = pd.Categorical(data.Gender)
data.Gender = data.Gender.cat.codes
data.JobRole = pd.Categorical(data.JobRole)
data.JobRole = data.JobRole.cat.codes
data.MaritalStatus = pd.Categorical(data.MaritalStatus)
data.MaritalStatus = data.MaritalStatus.cat.codes
data.Over18 = pd.Categorical(data.Over18)
data.Over18 = data.Over18.cat.codes
data.OverTime = pd.Categorical(data.OverTime)
data.OverTime = data.OverTime.cat.codes
data.head()

Age  Attrition  ...  YearsSinceLastPromotion  YearsWithCurrManager
0   41          1  ...                        0                     5
1   49          0  ...                        1                     7
2   37          1  ...                        0                     0
3   33          0  ...                        3                     0
4   27          0  ...                        2                     2

[5 rows x 35 columns]

In [ ]:
#Rechecking the data to make sure all features have numerical values
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
Age                         1470 non-null int64
Attrition                   1470 non-null int8
BusinessTravel              1470 non-null int8
DailyRate                   1470 non-null int64
Department                  1470 non-null int8
DistanceFromHome            1470 non-null int64
Education                   1470 non-null int64
EducationField              1470 non-null int8
EmployeeCount               1470 non-null int64
EmployeeNumber              1470 non-null int64
EnvironmentSatisfaction     1470 non-null int64
Gender                      1470 non-null int8
HourlyRate                  1470 non-null int64
JobInvolvement              1470 non-null int64
JobLevel                    1470 non-null int64
JobRole                     1470 non-null int8
JobSatisfaction             1470 non-null int64
MaritalStatus               1470 non-null int8
MonthlyIncome               1470 non

In [ ]:
# Dropping Infinites and NaN values if there are any
data.replace([np.inf, -np.inf], np.nan)
data.replace([np.inf, -np.inf],np.nan).dropna()

Age  Attrition  ...  YearsSinceLastPromotion  YearsWithCurrManager
0      41          1  ...                        0                     5
1      49          0  ...                        1                     7
2      37          1  ...                        0                     0
3      33          0  ...                        3                     0
4      27          0  ...                        2                     2
...   ...        ...  ...                      ...                   ...
1465   36          0  ...                        0                     3
1466   39          0  ...                        1                     7
1467   27          0  ...                        0                     3
1468   49          0  ...                        0                     8
1469   34          0  ...                        1                     2

[1470 rows x 35 columns]

In [ ]:
data['Attrition'].value_counts()

0    1233
1     237
Name: Attrition, dtype: int64

In [ ]:
#Downsample
data_majority = data[data.Attrition==0]
data_minority = data[data.Attrition==1]
data_majority_downsampled = resample(data_majority, replace= False, n_samples= len(data_minority), random_state=123) 
data_downsampled = pd.concat([data_majority_downsampled, data_minority])

As we see, we have imbalanced classes. We will use sampling method to balance the classes.

In [ ]:
# Check value_counts and remove any columns with singular values
single_cols = []
for col in data.columns:
    if len(data[col].unique())==1:
        print(col)
        single_cols.append(col)
# Now lets drop the columns
data.drop(single_cols, axis=1, inplace=True)

EmployeeCount
Over18
StandardHours


In [ ]:
# Separate input features (X) and target variable (y)
Y = data.Attrition
X = data.drop(['Attrition'], axis=1)

In [ ]:
# Spliting dataset into training and test data
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2)

In [ ]:
#Creating instance of Logistic Regression model and fiting data in it.
LogisticR = LogisticRegression(max_iter=10000)
LogisticR.fit(x_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=10000,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
#Prediction
y_pred=LogisticR.predict(x_test)

In [ ]:
from sklearn import metrics
ConfMatrix = metrics.confusion_matrix(y_test, y_pred)
ConfMatrix

array([[243,   6],
       [ 27,  18]])

Logstic model made 242 and 17 accurate predictions whereas 7 and 28 are inaccurate predictions

In [ ]:
#Evaluation of the model
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))
print("Precision:",metrics.precision_score(y_test, y_pred))
print("Recall:",metrics.recall_score(y_test, y_pred))

Accuracy: 0.8877551020408163
Precision: 0.75
Recall: 0.4


The Accuracy of our model is 88%. As per Precision, employee attrition rate would be 70%. If there are some employees who have quit their job in test data, our model can identify it 37% of the time.

*We* have implemented Logistic regression to classify values in 'Attrition' column from our dataset. We chose Attrition feature as it contains catagorical values 'Yes' and 'No' which depicts the attrition status of an employee. 
As the objective of our project is to predict employee attrition from given data, choosing Attrition column as feature will help in further analysis.


Logistic Regression computes the probability of an employee quitting his job. This model helps in solving our business problem; i.e. predicting whether an employee attrites or not.